File: backend/Services/Parsers/QuordleParser.cs
using System.Text.RegularExpressions;
using backend.Models;

namespace backend.Services.Parsers
{
    public class QuordleParser : ResultParser
    {
        private ILogger<QuordleParser> _logger;

        public QuordleParser(ILogger<QuordleParser> logger) : base(logger)
        {
            _logger = logger;
        }

        public override bool CountWinner => true;
        public override bool Default => true;
        private List<string> ScoreGroups = new [] { 1, 2, 3, 4 }
            .Select(i => "score" + i.ToString())
            .ToList();

        public override string GameName => "Quordle";
        public override bool GolfScoring => true;
        public override string? HelpText => null;

        protected override Regex Parser => new Regex($"Daily {GameName}[^\\d]+\\d+.*?[\\s\n\r]+{string.Join("[^\\d]*", ScoreGroups.Select(g => $"(?<{g}>[\\d|🟥])"))}");
        public override string Url => "https://www.merriam-webster.com/games/quordle/";

        protected override string GetCleanResult(string result, Match parserResults)
        {
            return result.Replace("m-w.com/games/quordle/", string.Empty).Trim();
        }

        public override string? GetScoreValue(DailyResult dailyResult)
        {
            if (dailyResult.Scores == null)
            {
                return null;
            }

            return "\"" + string.Join(",", Enumerable.Range(0, 4).ToArray()
                .Select((i) => {
                    if (dailyResult.Scores.Count <= i)
                    {
                        return "X";
                    }

                    return dailyResult.Scores[i].ToString();
                })) + "\"";
        }

        protected override DailyResult SetScore(DailyResult dailyResult, Match parserResults)
        {
            var scores = ScoreGroups
                .Where(g => parserResults.Groups.ContainsKey(g))
                .Select(g => parserResults.Groups[g].Value)
                .Select(v => int.TryParse(v, out var score) ? (int?)score : null)
                .Where(v => v.HasValue)
                .Select(v => v!.Value)
                .ToList();

            dailyResult.Scores = scores;

            return dailyResult;
        }
    }
}

File: lambdas/src/parsers/quordle-parser.ts
import { DailyResult } from '../models/daily-result';
import { ResultParser, getGroup, hasGroup } from './base/result-parser';

export class QuordleParser extends ResultParser {
  readonly countWinner = true;
  readonly gameName = 'Quordle';
  readonly golfScoring = true;
  readonly helpText = null;
  readonly url = 'https://www.merriam-webster.com/games/quordle/';
  
  private readonly scoreGroups = ['score1', 'score2', 'score3', 'score4'];
  
  // Note: C# uses [\d|🟥] which incorrectly includes literal pipe - fixed here to [\d🟥]
  protected readonly parser = new RegExp(
    `Daily ${this.gameName}[^\\d]+\\d+.*?[\\s\\n\\r]+` +
    this.scoreGroups.map(g => `(?<${g}>[\\d🟥])`).join('[^\\d]*')
  );
  
  protected getCleanResult(result: string, match: RegExpMatchArray): string {
    return result.replace('m-w.com/games/quordle/', '').trim();
  }
  
  getScoreValue(dailyResult: DailyResult): string | null {
    if (!dailyResult.scores) return null;
    
    const values = [0, 1, 2, 3].map(i => {
      if (dailyResult.scores && dailyResult.scores.length > i) {
        return dailyResult.scores[i].toString();
      }
      return 'X';
    });
    
    return `"${values.join(',')}"`;
  }
  
  protected setScore(dailyResult: DailyResult, match: RegExpMatchArray): DailyResult {
    const scores: number[] = [];
    
    for (const group of this.scoreGroups) {
      if (hasGroup(match, group)) {
        const value = getGroup(match, group);
        if (value) {
          const score = parseInt(value, 10);
          if (!isNaN(score)) {
            scores.push(score);
          }
        }
      }
    }
    
    if (scores.length > 0) {
      dailyResult.scores = scores;
    }
    
    return dailyResult;
  }
}
